refactor(ivy): correctly type class declarations

gkalpak opened the pull request:
Previously, several `ngtsc` and `ngcc` APIs dealing with class declaration nodes used inconsistent types. For example, some methods of the `DecoratorHandler` interface expected a `ts.Declaration` argument, but actual `DecoratorHandler` implementations specified a stricter `ts.ClassDeclaration` type.

As a result, the stricter methods would operate under the incorrect assumption that their arguments were of type `ts.ClassDeclaration`, while the actual arguments might be of different types (e.g. `ngcc` would call them with `ts.FunctionDeclaration` or `ts.VariableDeclaration` arguments, when compiling ES5 code).

Additionally, since we need those class declarations to be referenced in other parts of the program, `ngtsc`/`ngcc` had to either repeatedly check for `ts.isIdentifier(node.name)` or assume there was a `name` identifier and use `node.name!`. While this assumption happens to be true in the current implementation, working around type-checking is error-prone (e.g. the assumption might stop being true in the future).

This commit fixes this by introducing a new type to be used for such class declarations (`ts.Declaration & {name: ts.Identifier}`) and using it consistently throughput the code.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A


## What is the new behavior?


## Does this PR introduce a breaking change?

- [ ] Yes
- [ ] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information
Jira issue: [FW-1151][1]

[1]: https://angular-team.atlassian.net/browse/FW-1151

alfaproject commented on packages/compiler-cli/src/ngtsc/scope/src/util.ts:
```diff
@@ -88,6 +88,10 @@ export function extractDirectiveGuards(node: ts.Declaration, reflector: Reflecti
 }
 
+function isClassDeclaration(node: ts.Declaration): node is ClassDeclaration {
+  return ts.isClassDeclaration(node) && (node.name === undefined);
```
`=== undefined`?

gkalpak replied:
:grin: (It's a draft 😇)

petebacondarwin commented on packages/compiler-cli/src/ngcc/src/host/esm2015_host.ts:
```diff
@@ -392,6 +316,26 @@ export class Esm2015ReflectionHost extends TypeScriptReflectionHost implements N
+      return undefined;
+    }
+
+    return hasNameIdentifier(node) ? node : undefined;
```
Could be simplified

petebacondarwin replied:
```ts
return ts.isClassDeclaration(node) || ts.isClassExpression(node) && hasNameIdentifier(node) ?
    node :
    undefined;
```

gkalpak replied:
Actually, it should be `... (ts.isClassDeclaration(node) || ts.isClassExpression(node)) && hasNameIdentifier(node)...`. :grin:
I was holding back, because I thought you (and especially @alxhub) preferred the explicitness of separate `if` blocks.

petebacondarwin commented on packages/compiler-cli/src/ngcc/src/host/esm2015_host.ts:
```diff
@@ -80,88 +80,19 @@ export class Esm2015ReflectionHost extends TypeScriptReflectionHost implements N
    *
    * @throws if `declaration` does not resolve to a class declaration.
    */
   getMembersOfClass(clazz: ts.Declaration): ClassMember[] {
```
We ought to be able to move all such typings from `clazz: ts.Declaration` to the new `ClassDeclaration` type, I believe.

gkalpak replied:
Yeah, I was thinking about that, but (a) wasn't sure if these methods are only called with `ClassDeclaration` (and didn't bother finding out :grin:) and (b) I didn't want to make unnecessary changes (i.e. I only fixed the necessary signatures and then made necessary changes for typechecking to be happy).

Happy to go "all in" :grin:

petebacondarwin commented on packages/compiler-cli/src/ngcc/src/host/esm5_host.ts:
```diff
@@ -37,63 +37,28 @@ export class Esm5ReflectionHost extends Esm2015ReflectionHost {
 
-    const iifeBody = classSymbol.valueDeclaration.parent;
-    if (!iifeBody || !ts.isBlock(iifeBody)) return false;
+    const iifeBody = getIifeBody(classDeclaration);
```
:+1:

petebacondarwin commented on packages/compiler-cli/src/ngcc/src/rendering/esm5_renderer.ts:
```diff
@@ -23,21 +24,18 @@ export class Esm5Renderer extends EsmRenderer {
-    }
-    const parent = classSymbol.valueDeclaration && classSymbol.valueDeclaration.parent;
-    if (!parent || !ts.isBlock(parent)) {
+    const iifeBody = getIifeBody(compiledClass.declaration);
```
So is it now the case that the `compiledClass.declaration` is the outer declaration?

gkalpak replied:
Exactly.

petebacondarwin commented on packages/compiler-cli/src/ngcc/src/host/esm5_host.ts:
```diff
@@ -392,14 +445,75 @@ function getClassDeclarationFromInnerFunctionDeclaration(node: ts.Node): ts.Vari
+ * @param checker the TS program TypeChecker
+ * @returns the inner function declaration or `undefined` if it is not a "class".
+ */
+function getInnerFunctionDeclarationFromClassDeclaration(
```
Making functions like these into protected methods on the host would have two benefits:
1) we can avoid having to pass the `typechecker` as a parameter since it is a member of the host.
2) we can override it in a future `UmdReflectionHost` that might need slightly different behaviour.

gkalpak replied:
It didn't use the `TypeChecker` in an earlier implementation, but totally makes sense now.
<sub>(Personally, I would turn all standalone functions into methods (or move them to a `utils` file if they are generic and re-usable).)</sub>

petebacondarwin reviewed (commented):
I know this is still draft but as it might solve my problem I started to look at it.
I like the idea of a new `ClassDeclaration` type. This kind of abstraction fits well with our general approach to hiding the internals of the TS AST. I believe we should go a bit further and use it wherever we can to hide the use of `ts.ClassDeclaration`.
I also think that the code would be easier to decipher if we added new alias types for `InnerClassDeclaration` and `OuterClassDeclaration` (or perhaps just let the latter be `ClassDeclaration`).

gkalpak commented:
> I also think that the code would be easier to decipher if we added new alias types for `InnerClassDeclaration` and `OuterClassDeclaration` (or perhaps just let the latter be `ClassDeclaration`).

I briefly looked into it, but I didn't where how it would be useful. If it's an ES5 thing only, we can't use it outside of `Esm5ReflectionHost`, and many (most?) methods can receive either one, so 🤷‍♂️

gkalpak commented:
BTW, one optimization that can be done is memoizing [outer variable declaration <--> inner function declaration] pairs, so that we don't have to walk the AST in order to get from one to the other. I was going to do that, but maybe not worth it if this code is going to be refactored as part of #29119 anyway.

petebacondarwin commented on packages/compiler-cli/src/ngtsc/reflection/src/host.ts:
```diff
@@ -42,6 +42,28 @@ export interface Decorator {
+ * For `ReflectionHost` purposes, a class declaration should always have a `name` identifier,
+ * because we need to be able to reference it in other parts of the program.
+ */
+export type ClassDeclaration<T extends ts.Declaration = ts.Declaration> = T & {name: ts.Identifier};
```
I don't think it matters but `ts.Declaration` could be replaced with `ts.NamedDeclaration` right?

gkalpak replied:
It probably could, but it wouldn't make any difference (since we are explicitly typing `name` anyway).
Notice that (unintuitively?) `ts.NamedDeclaration` is not guaranteed to have a `name` and even if it has it is not necessarily a `ts.Identifier`:

```ts
interface NamedDeclaration extends Declaration {
  name?: Identifier | StringLiteralLike | NumericLiteral | ComputedPropertyName | BindingPattern;
}
```

petebacondarwin commented on packages/compiler-cli/src/ngtsc/transform/src/compilation.ts:
```diff
@@ -291,8 +291,8 @@ export class IvyCompilation {
     // Look to see whether the original node was analyzed. If not, there's nothing to do.
-    const original = ts.getOriginalNode(node) as ts.Declaration;
-    if (!this.ivyClasses.has(original)) {
+    const original = ts.getOriginalNode(node) as typeof node;
```
Does this make any difference? `typeof node === ts.Declaration` right?

gkalpak replied:
Yeah, it doesn;t make a difference (just making the code DRYer :grin:).

petebacondarwin commented on packages/compiler-cli/src/ngcc/test/host/util.ts:
```diff
@@ -29,3 +29,18 @@ export function expectTypeValueReferencesForParameters(
+export function isNamedVariableDeclaration(node: ts.Node):
+    node is ClassDeclaration<ts.VariableDeclaration> {
+  return ts.isVariableDeclaration(node) && (node.name !== undefined);
+}
```
Love these helpers :-) Are they general enough that we should put them in with the definitions of `ClassDeclaration` etc?

gkalpak replied:
We could definitely do that, but I didn't want to polute the source code with stuff that isn't used. Two of the three mostly make sense for `ngcc` (which deals with non-TS code), so putting them in ngtsc (where `ClassDeclaration` is defined) didn't feel right.
I thought a little code duplication (of `isNamedClassDeclaration()`) was fine in this case. But happy to move them around, if think it is worth it.

petebacondarwin commented on packages/compiler-cli/src/ngcc/src/host/esm5_host.ts:
```diff
@@ -62,38 +62,39 @@ export class Esm5ReflectionHost extends Esm2015ReflectionHost {
-      if (!iifeBody) return undefined;
+  getClassDeclaration(node: ts.Node): ClassDeclaration|undefined {
+    const superDeclaration = super.getClassDeclaration(node);
+    if (superDeclaration) return superDeclaration;
```
I don't think there is much point in making a call to `super.getClassDeclaration()` from ES5 code, since the super version will only look for `ClassDeclaration` and `ClassExpression` nodes, of which there should be none in ES5. Right?

petebacondarwin replied:
Oh I see that we might hit a typings file... OK then.

petebacondarwin commented on packages/compiler-cli/src/ngcc/src/host/esm2015_host.ts:
```diff
@@ -54,6 +54,37 @@ export class Esm2015ReflectionHost extends TypeScriptReflectionHost implements N
+  getClassDeclaration(node: ts.Node): ClassDeclaration|undefined {
+    if (ts.isVariableDeclaration(node) && node.initializer) {
+      node = node.initializer;
+    }
```
I don't know how I feel about this any more. I know you just copied it from the previous implementation.
I wonder under what scenarios the class is defined as `const SomeAlias = class SomeClass {};`?
And in those scenarios do we care more about the `SomeClass` than the `SomeAlias`?

petebacondarwin replied:
Oh I see that we account for this further down when searching for static members:

```ts
if (ts.isVariableDeclaration(symbol.valueDeclaration.parent)) {
```

Perhaps #29119 will tidy this up...

petebacondarwin commented on packages/compiler-cli/src/ngcc/src/host/esm2015_host.ts:
```diff
@@ -54,6 +54,37 @@ export class Esm2015ReflectionHost extends TypeScriptReflectionHost implements N
+   * @param node the node that represents the class whose declaration we are finding.
+   * @returns the declaration of the class or `undefined` if it is not a "class".
+   */
+  getClassDeclaration(node: ts.Node): ClassDeclaration|undefined {
```
AFAICT the only reason for making this `public` is so that it can be unit tested. Right?

gkalpak replied:
Right :disappointed:

petebacondarwin commented on packages/compiler-cli/src/ngcc/src/host/esm2015_host.ts:
```diff
@@ -493,6 +439,87 @@ export class Esm2015ReflectionHost extends TypeScriptReflectionHost implements N
+          }
+        });
+      }
+    }
```
Why we didn't use a little helper function for these three blocks of code, I'll never know...

gkalpak replied:
It's never too late :stuck_out_tongue:

petebacondarwin commented on packages/compiler-cli/src/ngcc/src/host/esm5_host.ts:
```diff
@@ -37,9 +37,7 @@ export class Esm5ReflectionHost extends Esm2015ReflectionHost {
-  isClass(node: ts.Node): node is ClassDeclaration {
-    return super.isClass(node) || !!this.getClassSymbol(node);
-  }
+  isClass(node: ts.Node): node is ClassDeclaration { return !!this.getClassDeclaration(node); }
```
`Esm2015ReflectionHost` does not implement it own `isClass()` but I wonder whether it ought to?
The `Esm2015ReflectionHost.getClassDeclaration()` method may return a `ts.ClassExpression` (i.e. a valid class) but `TypeScriptReflectionHost.isClass()` only returns true for `ts.ClassDeclaration`.

gkalpak replied:
Good point :thinking:
`getClassDeclaration()` is used for `getClassSymbol()`, but `isClass()` would indeed not return `true` for all cases that `getClassDeclaration()` can handle.

I will try moving `isClass()` to `Esm2015ReflectionHost` and see if anything break :stuck_out_tongue:

JoostK replied:
@petebacondarwin @gkalpak With #29119 the ESM2015 host *will* have its own `isClass()`. I coordinated with George last week that some follow-up work to streamline the two hosts was deemed the way to go.

I would propose to go forward with this PR without touching the ES2015 stuff to avoid conflicts with #29119 at this point, and align the two hosts in #29119 or even in a new PR altogether.

gkalpak replied:
Works for me. I added a TODO item in #29119's description, so we don't forget.

petebacondarwin commented on packages/compiler-cli/src/ngcc/src/host/esm5_host.ts:
```diff
@@ -49,11 +47,13 @@ export class Esm5ReflectionHost extends Esm2015ReflectionHost {
+    if (super.hasBaseClass(clazz)) return true;
+
+    const classDeclaration = this.getClassDeclaration(clazz);
+    if (!classDeclaration) return false;
```
Could be simplified to

```ts
if (!this.isClass(clazz)) return false;
```

gkalpak replied:
We are using `classDeclaration` below, so we need `getClassDeclaration()` anyway.

petebacondarwin replied:
Oops - yes, I didn't spot that.

petebacondarwin commented on packages/compiler-cli/src/ngcc/src/host/esm5_host.ts:
```diff
@@ -62,38 +62,39 @@ export class Esm5ReflectionHost extends Esm2015ReflectionHost {
+    if (!this.getInnerFunctionDeclarationFromClassDeclaration(node)) return undefined;
 
-    const outerClassNode = getClassDeclarationFromInnerFunctionDeclaration(node);
+    const node_ = node as ts.VariableDeclaration;
```
We should try to avoid casts like this... Instead we can use a type guard, which will also short-circuit the call to `getInnerFunctionDeclarationFromClassDeclaration()` if it fails:

```ts
if (!ts.isVariableDeclaration(node) || !this. getInnerFunctionDeclarationFromClassDeclaration(node)) return undefined;
```

or even more concise (though perhaps going too far)

```ts
return ts.isVariableDeclaration(node) && hasNameIdentifier(node) && this.getInnerFunctionDeclarationFromClassDeclaration(node)) ? node : undefined;
```

petebacondarwin replied:
Actually the second version is the way to go, as it doubly-short-circuits the more complex call.

petebacondarwin replied:
Or even better if those helpers were made more widely available:

```ts
return isNamedVariableDeclaration(node) && this.getInnerFunctionDeclarationFromClassDeclaration(node)) ? node : undefined;
```

gkalpak replied:
Done.

petebacondarwin commented on packages/compiler-cli/src/ngcc/src/host/decorated_class.ts:
```diff
@@ -22,5 +22,5 @@ export class DecoratedClass {
    */
   constructor(
-      public name: string, public declaration: ts.Declaration, public decorators: Decorator[], ) {}
+      public name: string, public declaration: ClassDeclaration, public decorators: Decorator[]) {}
```
We should update the docs for this parameter to clarify that in Esm5 the `declaration` is always the "outer" function.

gkalpak replied:
Done.

petebacondarwin commented on packages/compiler-cli/src/ngcc/test/rendering/esm5_renderer_spec.ts:
```diff
@@ -258,28 +258,16 @@ SOME DEFINITION TEXT
-      const hostSpy = spyOn(host, 'getClassSymbol').and.returnValue(null);
-      expect(() => renderer.addDefinitions(output, badSymbol, 'SOME DEFINITION TEXT'))
-          .toThrowError('Compiled class does not have a valid symbol: BadSymbol in /some/file.js');
-
```
Why remove this test?

gkalpak replied:
Because the relavant part was removed from the `addDefinitions()` method. It no longer uses `host.getClassSymbol()`.

petebacondarwin commented on packages/compiler-cli/src/ngcc/test/host/esm5_host_spec.ts:
```diff
@@ -1384,21 +1391,12 @@ describe('Esm5ReflectionHost', () => {
+      const host = new Esm5ReflectionHost(false, program.getTypeChecker());
+      const node = getDeclaration(
+          program, SIMPLE_ES2015_CLASS_FILE.name, 'EmptyClass', isNamedClassDeclaration);
+      expect(host.getClassSymbol(node)).toBeDefined();
```
We should probably now test that the return value is what we expect (not just defined).

gkalpak replied:
Done.

petebacondarwin reviewed (changes requested):
This is looking great! I made a couple of suggestions which do not block approval.

But one thing that worries me is the following method on `Esm5ReflectionHost`, which does not seem to have been updated:

```ts
  protected getStatementsForClass(classSymbol: ClassSymbol): ts.Statement[] {
    const classDeclarationParent = classSymbol.valueDeclaration.parent;
    return ts.isBlock(classDeclarationParent) ? Array.from(classDeclarationParent.statements) : [];
  }
}
```

Note that now that `classSymbol` should always refer to the outer declaration, this method should return the top level statement from the file, rather than the IIFE statements that we actually need to check.

But if that is so, then I don't understand why the unit tests are not failing 🤔

gkalpak commented:
The reason this works is that this method is always called transitively through methods that pass the inner function declaration:
- [getDecoratorsOfSymbol()][1]
- [getMembersOfClass()][2]
- [getConstructorParamInfo()][3]

We I should look into making the types stricter and have two kinds of "ClassDefinition" (as you suggested in https://github.com/angular/angular/pull/29209#pullrequestreview-213244117). Something like having a "PublicClassDeclaration" (for what the rest of the program sees as the class declaration) and a "ClassDeclarationImplementation" (which has the properties/methods/decorators).

This would give us more type-safety. My only concern is that (in order for them to be useful) the types need to leak to all `ReflectionHost`s, even though they only make sense in ES5. WDYT?

[1]: https://github.com/angular/angular/blob/cf9abf9e63afbfde9e2fbe7912f3010f59084cb5/packages/compiler-cli/src/ngcc/src/host/esm5_host.ts#L285-L292
[2]: https://github.com/angular/angular/blob/cf9abf9e63afbfde9e2fbe7912f3010f59084cb5/packages/compiler-cli/src/ngcc/src/host/esm5_host.ts#L162-L173
[3]: https://github.com/angular/angular/blob/cf9abf9e63afbfde9e2fbe7912f3010f59084cb5/packages/compiler-cli/src/ngcc/src/host/esm5_host.ts#L275-L283

petebacondarwin commented:
Actually there is already a precedent for this in `ClassMember` which has an `implementation` property. Perhaps `ClassDeclaration` should be an interface something like:

```ts
interface ClassDeclaration {
  name: ts.Identifier;
  declaration: ts.Declaration;
  implementation?: ts.Declaration;
}
```

This would also all us to cache the inner <=> outer mapping to improve performance...

mary-poppins commented:
You can preview 6846e9c at https://pr29209-6846e9c.ngbuilds.io/.
You can preview 2112ad8 at https://pr29209-2112ad8.ngbuilds.io/.
You can preview 0d88d10 at https://pr29209-0d88d10.ngbuilds.io/.
You can preview dc19808 at https://pr29209-dc19808.ngbuilds.io/.
You can preview 5fba061 at https://pr29209-5fba061.ngbuilds.io/.
You can preview f530bd0 at https://pr29209-f530bd0.ngbuilds.io/.
You can preview c853370 at https://pr29209-c853370.ngbuilds.io/.
You can preview a086804 at https://pr29209-a086804.ngbuilds.io/.
You can preview a4721f6 at https://pr29209-a4721f6.ngbuilds.io/.
You can preview 9db9df5 at https://pr29209-9db9df5.ngbuilds.io/.
You can preview cf9abf9 at https://pr29209-cf9abf9.ngbuilds.io/.

mary-poppins commented:
You can preview 8a91e18 at https://pr29209-8a91e18.ngbuilds.io/.

mary-poppins commented:
You can preview 0653416 at https://pr29209-0653416.ngbuilds.io/.

mary-poppins commented:
You can preview 17851d2 at https://pr29209-17851d2.ngbuilds.io/.

mary-poppins commented:
You can preview 41be5c9 at https://pr29209-41be5c9.ngbuilds.io/.

gkalpak commented:
OK, I think this is good to go. @petebacondarwin, PTAL.
The following refactorings will be done in another PR (probably after #29119):
- Implement `getClassDeclaration()`-based `isClass()` in `Esm2015ReflectionHost` (https://github.com/angular/angular/pull/29209#discussion_r265914647).
- Implement `ClassDeclaration` as a wrapper object containing `declaration` and `implementation` (https://github.com/angular/angular/pull/29209#issuecomment-473269514).

mary-poppins commented:
You can preview 14dc4dc at https://pr29209-14dc4dc.ngbuilds.io/.

petebacondarwin commented on packages/compiler-cli/src/ngcc/src/host/esm5_host.ts:
```diff
@@ -42,12 +42,14 @@ export class Esm5ReflectionHost extends Esm2015ReflectionHost {
 
   /**
-   * Determines whether the given declaration has a base class.
+   * Determines whether the given declaration, which should be a class, has a base class.
```
I would be tempted to put "class" in quotes here, since in ESM5 it is not literally a `class`.

gkalpak committed:
refactor(ivy): correctly type class declarations in `ngtsc`/`ngcc`

Previously, several `ngtsc` and `ngcc` APIs dealing with class
declaration nodes used inconsistent types. For example, some methods of
the `DecoratorHandler` interface expected a `ts.Declaration` argument,
but actual `DecoratorHandler` implementations specified a stricter
`ts.ClassDeclaration` type.

As a result, the stricter methods would operate under the incorrect
assumption that their arguments were of type `ts.ClassDeclaration`,
while the actual arguments might be of different types (e.g. `ngcc`
would call them with `ts.FunctionDeclaration` or
`ts.VariableDeclaration` arguments, when compiling ES5 code).

Additionally, since we need those class declarations to be referenced in
other parts of the program, `ngtsc`/`ngcc` had to either repeatedly
check for `ts.isIdentifier(node.name)` or assume there was a `name`
identifier and use `node.name!`. While this assumption happens to be
true in the current implementation, working around type-checking is
error-prone (e.g. the assumption might stop being true in the future).

This commit fixes this by introducing a new type to be used for such
class declarations (`ts.Declaration & {name: ts.Identifier}`) and using
it consistently throughput the code.

gkalpak committed:
fixup! refactor(ivy): correctly type class declarations in `ngtsc`/`ngcc`

gkalpak committed:
fix(ivy): handle class declarations consistently in ES5 code

gkalpak committed:
refactor(ivy): use `ClassDeclaration` in more `ReflectionHost` methods

gkalpak committed:
fixup! refactor(ivy): use `ClassDeclaration` in more `ReflectionHost` methods

gkalpak committed:
fixup! fix(ivy): handle class declarations consistently in ES5 code

mary-poppins commented:
You can preview c170e47 at https://pr29209-c170e47.ngbuilds.io/.

petebacondarwin commented on packages/compiler-cli/src/ngcc/src/host/esm2015_host.ts:
```diff
@@ -493,6 +439,87 @@ export class Esm2015ReflectionHost extends TypeScriptReflectionHost implements N
+   * Examine a symbol which should be of a class, and return metadata about its members.
+   *
+   * @param symbol the `ClassSymbol` representing the class over which to reflect. If the source is
+   * in
```
😱

petebacondarwin commented on packages/compiler-cli/src/ngcc/src/host/esm2015_host.ts:
```diff
@@ -493,6 +439,87 @@ export class Esm2015ReflectionHost extends TypeScriptReflectionHost implements N
+   * @param symbol the `ClassSymbol` representing the class over which to reflect. If the source is
+   * in
+   * ES2015 format, this will be a `ts.ClassDeclaration` node. If the source is in ES5 format, this
+   * might be a `ts.FunctionDeclaration` as classes in ES5 are downleveled to constructor functions.
```
I think the docs for this param are not accurate. Perhaps to avoid getting it wrong here, and having to keep it in sync, we should just simplify to:

```
@param symbol the `ClassSymbol` representing the class over which to reflect.
```

Then we should make sure that `getClassSymbol()` and/or `ClassSymbol` are well documented for the different cases in ES2015 and ES5.

petebacondarwin commented on packages/compiler-cli/src/ngcc/src/host/esm5_host.ts:
```diff
@@ -62,38 +62,39 @@ export class Esm5ReflectionHost extends Esm2015ReflectionHost {
-   * @param node the top level declaration that represents an exported class or the function
-   *     expression inside the IIFE.
-   * @returns the symbol for the node or `undefined` if it is not a "class" or has no symbol.
+   * The returned declaration is either the class declaration (from the declaration file) or the
```
I tend to prefer "typings file" to "declaration file". Declaration is a heavily overloaded term in this code :-)
And in the bullet above.

petebacondarwin commented on packages/compiler-cli/src/ngcc/src/host/decorated_class.ts:
```diff
@@ -17,7 +17,10 @@ export class DecoratedClass {
+   * @param declaration The TypeScript AST node where this class is declared. In ES5 code, where a
+   * class can be represented by both a variable declaration and a function declaration (inside an
+   * IIFE), `declaration` will always refer to the outer variable declaration, which represents the
+   * class to the rest of the program.
```
Also, worth stating that in ES2015 a class can be represented by a ClassExpression or a VariableDeclaration in some cases:

```ts
let MyClass = _MyClass = class MyClass {};
```

gkalpak replied:
This info is included in the description of the `ClassDeclaration` type. The note here is about a special case in ES5 wrt to `DecoratedClass` (i.e. that we only pass the outer variable declaration and not other types of `ClassDeclaration`s (e.g. the inner function)).

(This will also be obsolete once we implement the `ClassDeclaration` wrapper (with `implementation`).)

gkalpak committed:
fixup! refactor(ivy): correctly type class declarations in `ngtsc`/`ngcc`

gkalpak committed:
fixup! refactor(ivy): use `ClassDeclaration` in more `ReflectionHost` methods

gkalpak committed:
fixup! fix(ivy): handle class declarations consistently in ES5 code

mary-poppins commented:
You can preview 322e6aa at https://pr29209-322e6aa.ngbuilds.io/.

gkalpak commented:
merge-assistance: The `material-unit-tests` failures are unrelated to this PR (and will be fixed by #29416).

gkalpak committed:
refactor(ivy): remove unused code from `TypeCheckContext`

gkalpak committed:
refactor(ivy): implement `DtsModuleScopeResolver` from `MetadataDtsModuleScopeResolver`

mary-poppins commented:
You can preview 7ba59e1 at https://pr29209-7ba59e1.ngbuilds.io/.
